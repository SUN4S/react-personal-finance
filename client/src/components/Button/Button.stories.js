import { Button } from "./Button";

export default {
  title: "Button",
  component: Button,
};

export const Primary_Light = () => (
  <html data-theme="light">
    <Button type="button" text="Press me!" class="primaryBtn" />
  </html>
);

export const Primary_Dark = () => (
  <html data-theme="dark">
    <Button type="button" text="Press me!" class="primaryBtn" />
  </html>
);

export const Primary_Loading = () => (
  <html data-theme="dark">
    <Button
      type="button"
      text="Press me!"
      class="primaryBtn"
      loading
      disabled
    />
  </html>
);

export const Secondary_Light = () => (
  <html data-theme="light">
    <Button type="button" text="Press me!" class="secondaryBtn" />
  </html>
);

export const Secondary_Dark = () => (
  <html data-theme="dark">
    <Button type="button" text="Press me!" class="secondaryBtn" />
  </html>
);

export const Secondary_Loading = () => (
  <html data-theme="dark">
    <Button
      type="button"
      text="Press me!"
      class="secondaryBtn"
      loading
      disabled
    />
  </html>
);

export const Gray_light = () => (
  <html data-theme="light">
    <Button type="button" text="Press me!" class="grayBtn" />
  </html>
);

export const Gray_Light_Loading = () => (
  <html data-theme="light">
    <Button type="button" text="Press me!" class="grayBtn" loading disabled />
  </html>
);

export const Gray_Dark = () => (
  <html data-theme="dark">
    <Button type="button" text="Press me!" class="grayBtn" />
  </html>
);

export const Gray_Dark_Loading = () => (
  <html data-theme="dark">
    <Button type="button" text="Press me!" class="grayBtn" loading disabled />
  </html>
);
