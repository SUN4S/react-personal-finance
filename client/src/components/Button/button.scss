button {
  position: relative;
  overflow: hidden;
  transition: background 400ms;
  outline: 0;
  border: 0;
  cursor: pointer;
  font-size: 16px;
  border-radius: 3px;
  height: 40px;
  width: 100%;
  cursor: pointer;

  .loader {
    width: 20px;
    height: 20px;

    &::before {
      top: 2px;
      display: inline-block;
      position: relative;
      content: "";
      height: 20px;
      width: 20px;
      border-radius: 100px;
      border: 2px solid var(--text-secondary);
      border-top: 2px solid var(--primary);
      animation: loader 1s infinite linear;

      @keyframes loader {
        from {
          transform: rotateZ(0deg);
        }
        to {
          transform: rotateZ(360deg);
        }
      }
    }
  }
}

span.ripple {
  position: absolute; /* The absolute position we mentioned earlier */
  border-radius: 0%;
  transform: scaleX(0);
  animation: ripple 500ms linear;
  background-color: rgba(255, 255, 255, 0.7);
  border-radius: 600px;

  @keyframes ripple {
    to {
      transform: scaleX(4);
      opacity: 0;
    }
  }
}

.modalBtn {
  font-size: 16px;
  border-radius: 3px;
  height: 40px;
  width: 100%;
  margin-top: 24px;
  color: var(--text-primary);
  background-color: var(--primary);
  cursor: pointer;
}

.primaryBtn {
  color: rgb(241, 245, 249);
  background-color: var(--primary);
}

.secondaryBtn {
  color: rgb(241, 245, 249);
  background-color: var(--secondary);
}

.grayBtn {
  color: var(--text-primary);
  background-color: var(--sidebar-bg);
}

.dangerBtn {
  color: var(--text-primary);
  background-color: var(--alert);
}

.loadingBtn {
  height: 40px;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 16px;
  border-radius: 3px;
  margin-top: 32px;
  color: var(--text-primary);
  background-color: var(--sidebar-bg);
  cursor: pointer;

  &::before {
    display: inline-block;
    position: relative;
    content: "";
    height: 20px;
    width: 20px;
    border-radius: 100px;
    border: 2px solid var(--text-secondary);
    border-top: 2px solid var(--primary);
    animation: loader 1s infinite linear;

    @keyframes loader {
      from {
        transform: rotateZ(0deg);
      }
      to {
        transform: rotateZ(360deg);
      }
    }
  }
}
