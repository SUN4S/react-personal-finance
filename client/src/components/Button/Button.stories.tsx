import { Meta, Story } from "@storybook/react/types-6-0";

import { Button } from "./Button";

interface TemplateProps {
  theme: String;
  type: "button" | "submit" | "reset" | undefined;
  text: string;
  class: string;
  action?: Function;
  disabled: boolean;
  loading: boolean;
}

export default {
  title: "Components/Button",
  component: Button,
  argTypes: {
    loading: {
      options: [true, false],
      control: { type: "boolean" },
    },
    disabled: {
      options: [true, false],
      control: { type: "boolean" },
    },
  },
} as Meta;

const Template: Story<TemplateProps> = (args) => {
  return (
    <html data-theme={args.theme}>
      <Button {...args} />
    </html>
  );
};

export const Primary_Light = Template.bind({});
Primary_Light.args = {
  theme: "light",
  type: "button",
  text: "Click Me!",
  class: "primaryBtn",
  disabled: false,
  loading: false,
};

export const Primary_Dark = Template.bind({});
Primary_Dark.args = {
  theme: "dark",
  type: "button",
  text: "Click Me!",
  class: "primaryBtn",
  disabled: false,
  loading: false,
};

export const Secondary_Light = Template.bind({});
Secondary_Light.args = {
  theme: "light",
  type: "button",
  text: "Click Me!",
  class: "secondaryBtn",
  disabled: false,
  loading: false,
};

export const Secondary_Dark = Template.bind({});
Secondary_Dark.args = {
  theme: "dark",
  type: "button",
  text: "Click Me!",
  class: "secondaryBtn",
  disabled: false,
  loading: false,
};

export const Gray_Light = Template.bind({});
Gray_Light.args = {
  theme: "light",
  type: "button",
  text: "Click Me!",
  class: "grayBtn",
  disabled: false,
  loading: false,
};

export const Gray_Dark = Template.bind({});
Gray_Dark.args = {
  theme: "dark",
  type: "button",
  text: "Click Me!",
  class: "grayBtn",
  disabled: false,
  loading: false,
};
