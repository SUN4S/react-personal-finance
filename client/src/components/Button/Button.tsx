import "./Button.scss";

import React from "react";

// typescript defined button props
// 3 Required and 3 situational props
export type ButtonProps = {
  type: "button" | "submit" | "reset" | undefined;
  text: string;
  class: string;
  action?: Function;
  disabled?: boolean;
  loading?: boolean;
};

// Button prop with a ripple animation
export const Button = (props: ButtonProps) => {
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    // If action props exists, execute it
    // this is here because some components need to dispatch actions
    props.action && props.action();

    const button = e.currentTarget;
    // Modal breaks animation, so check is required
    // explanation: buttons are rendered on modal, which is pushed to the RIGHt off-screen
    // this causes ripple animation to be rendere off-screen aswell
    const modalElement = document.querySelector(".Modal")!;
    let modalWidth = 0;
    if (modalElement) {
      const modalStyle = getComputedStyle(modalElement).width;
      modalWidth = Number(modalStyle.replace(/\D/g, ""));
    }
    const checkIfModal = props.class.split(" ").includes("modalBtn")
      ? window.innerWidth - modalWidth
      : 0;

    // Hence forward are the ripple animation calculations
    const circle = document.createElement("span");
    const diameter = Math.max(button.clientWidth, button.clientHeight);
    const radius = diameter / 2;
    circle.style.width = circle.style.height = `${diameter}px`;
    circle.style.left = `${
      e.clientX - (button.offsetLeft + radius) - checkIfModal
    }px`;
    circle.style.top = `${e.clientY - (button.offsetTop + radius)}px`;
    circle.classList.add("ripple");
    const ripple = button.getElementsByClassName("ripple")[0];

    if (ripple) {
      ripple.remove();
    }
    button.appendChild(circle);
  };

  return (
    <button
      type={props.type}
      className={props.class}
      onClick={handleClick}
      disabled={props.disabled}
      data-testid="button"
    >
      {props.loading ? (
        <span className="loader" data-testid="buttonLoader"></span>
      ) : (
        props.text
      )}
    </button>
  );
};
