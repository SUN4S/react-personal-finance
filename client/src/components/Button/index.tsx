import "./button.scss";

import React from "react";

export type ButtonProps = {
  type: "button" | "submit" | "reset" | undefined;
  text: string;
  class: string;
  action?: Function;
  disabled?: boolean;
};

export const Button = (props: ButtonProps) => {
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    props.action && props.action();
    const button = e.currentTarget;
    // Modal breaks animation, so check is required
    const modalElement = document.querySelector(".Modal")!;
    let modalWidth = 0;
    if (modalElement) {
      const modalStyle = getComputedStyle(modalElement).width;
      modalWidth = Number(modalStyle.replace(/\D/g, ""));
    }
    const checkIfModal = props.class.split(" ").includes("modalBtn")
      ? window.innerWidth - modalWidth
      : 0;

    const circle = document.createElement("span");
    const diameter = Math.max(button.clientWidth, button.clientHeight);
    const radius = diameter / 2;
    circle.style.width = circle.style.height = `${diameter}px`;
    circle.style.left = `${
      e.clientX - (button.offsetLeft + radius) - checkIfModal
    }px`;
    circle.style.top = `${e.clientY - (button.offsetTop + radius)}px`;
    circle.classList.add("ripple");
    const ripple = button.getElementsByClassName("ripple")[0];

    if (ripple) {
      ripple.remove();
    }
    button.appendChild(circle);
  };

  return (
    <button
      type={props.type}
      className={props.class}
      onClick={handleClick}
      disabled={props.disabled}
    >
      {props.text}
    </button>
  );
};
