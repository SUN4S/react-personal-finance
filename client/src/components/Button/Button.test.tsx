import { cleanup, fireEvent, render } from "@testing-library/react";

import { Button } from "./Button";
import renderer from "react-test-renderer";

describe("Button", () => {
  afterEach(cleanup);

  it("Renders Button", () => {
    const { getByTestId } = render(
      <Button type="button" text="Press Me!" class="primaryBtn" />
    );
    expect(getByTestId("button")).toBeTruthy();
  });

  it("Button Fires handleclick", () => {
    const mockFunction = jest.fn();
    const { getByTestId } = render(
      <Button
        type="button"
        text="Press Me!"
        class="primaryBtn"
        action={mockFunction}
      />
    );
    fireEvent.click(getByTestId("button"));
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });

  it("Button Loading", () => {
    const { getByTestId } = render(
      <Button type="button" text="Press Me!" class="primaryBtn" loading />
    );
    expect(getByTestId("buttonLoader")).toBeTruthy();
  });

  it("Button Matches Snapshot", () => {
    const tree = renderer.create(
      <Button type="button" text="Press Me!" class="primaryBtn" />
    );
    expect(tree).toMatchSnapshot();
  });

  it("Button Matches Loading Snapshot", () => {
    const tree = renderer.create(
      <Button type="button" text="Press Me!" class="primaryBtn" loading />
    );
    expect(tree).toMatchSnapshot();
  });
});
