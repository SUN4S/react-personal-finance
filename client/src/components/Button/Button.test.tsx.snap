// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Matches Loading Snapshot 1`] = `
<button
  className="primaryBtn"
  data-testid="button"
  onClick={[Function]}
  type="button"
>
  <span
    className="loader"
    data-testid="buttonLoader"
  />
</button>
`;

exports[`Button Matches Snapshot 1`] = `
<button
  className="primaryBtn"
  data-testid="button"
  onClick={[Function]}
  type="button"
>
  Press Me!
</button>
`;
